txpipe/twopoint.py

```diff
@@ -1092,6 +1114,235 @@ def calculate_pos_pos(self, i, j):
+        ext_list = np.arange(len( self.ext_list ))
+
+        # For shear-ext
+        if self.config["do_shear_ext"]:
+            k = SHEAR_EXT
+            for i in source_list:
+                for j in ext_list:
+                    calcs.append((i, j, k))
+
+        if self.config["do_pos_ext"]:
+            k = POS_EXT
+            for i in lens_list:
+                for j in ext_list:
+                    calcs.append((i, j, k))
+
+        if self.rank == 0:
+            print(f"Running {len(calcs)} calculations: {calcs}")
+
+        return calcs
+
+    def calculate_pos_ext(self, i, j):
+        import treecorr
+
+        cat_i = self.get_density_map(i)
+
+        cat_j = self.get_ext_map(j)
+
+        if self.rank == 0:
+            print(
+                f"Calculating position-external bin pair ({i}, {j})"
+            )
+
+        t1 = perf_counter()
+
+        kk = treecorr.KKCorrelation(self.config)
+        kk.process(cat_i, cat_j, comm=self.comm, low_mem=self.config["low_mem"])
+
+        return kk
+
+
+    def calculate_shear_ext(self, i, j):
+        import treecorr
+
+        cat_i = self.get_shear_map(i)
+
+        cat_j = self.get_ext_map(j)
+
+        if self.rank == 0:
+            print(
+                f"Calculating shear-external bin pair ({i},{j})."
+            )
+
+        kg = treecorr.KGCorrelation(self.config)
+        t1 = perf_counter()
+        kg.process(cat_j, cat_i, comm=self.comm, low_mem=self.config["low_mem"])
+
+        return kg
+
+    def get_ext_map(self, i):
+        """
+        get the ith external map (e.g. SP map) from the directory specified in the config 
+        """
+        import treecorr
+        import healsparse
+        import healpy as hp 
+        
+        # Read the mask to get fracdet weights
+        with self.open_input("mask", wrapper=True) as f:
+            info = f.read_map_info('mask')
+            mask, pix, nside = f.read_healpix('mask', return_all=True)
+
+        # open the input healsparse map
+        sys_files = self.get_ext_list()
+        sys_file = sys_files[i]
+        m = healsparse.HealSparseMap.read(sys_file)
+        m = m.degrade(nside)
+        if info['nest'] == True:
+            map_d = m[pix]
+        else:
+            map_d = m[hp.ring2nest(nside, pix)] 
+
+        #convert SP map to an overdensity
+        mean_data = np.average(map_d, weights=mask[pix] )
+        map_d = map_d/mean_data - 1.
+
+        scheme = choose_pixelization(**info) 
+        ra_pix, dec_pix = scheme.pix2ang(pix)
+
+        # Load and calibrate the appropriate bin data
+        cat = treecorr.Catalog(
+            ra=ra_pix,
+            dec=dec_pix,
+            w=mask[pix], #weight pixels by their fracdet
+            k=map_d,
+            ra_units="degree",
+            dec_units="degree",
+            patch_centers=self.get_input("patch_centers"),
+        )
+        return cat
+
+
+    def write_output(self, source_list, lens_list, meta, results):
```

Comment:
I think you need to provide something more to SACC to get it to actually save your file no?

Reply:
seems to be saving for me. I'll add the example to demonstrate

Reply:
example added